scylla/src/transport/execution_profile.rs

```diff
@@ -0,0 +1,154 @@
+            speculative_execution_policy: self
+                .speculative_execution_policy
+                .unwrap_or_else(defaults::speculative_execution_policy),
+        }))
+    }
+}
+
+/// A profile that groups configurable options regarding query execution.
+///
+/// Execution profile is immutable as such, but the driver implements double indirection of form:
+/// query/Session -> ExecutionProfileHandle -> ExecutionProfile
+/// which enables on-fly changing the actual profile associated with all entities (query/Session)
+/// by the same handle.
+#[derive(Debug)]
+pub struct ExecutionProfile(pub(crate) Arc<ExecutionProfileInner>);
+
+#[derive(Debug)]
+pub(crate) struct ExecutionProfileInner {
+    pub(crate) request_timeout: Option<Duration>,
+    pub(crate) consistency: Consistency,
+    pub(crate) serial_consistency: Option<SerialConsistency>,
+    // ContactPointList blacklist_;
+    // DcList blacklist_dc_;
+    // LatencyAwarePolicy::Settings latency_aware_routing_settings_;
+    // ContactPointList whitelist_;
+    // DcList whitelist_dc_;
+    pub(crate) load_balancing_policy: Arc<dyn LoadBalancingPolicy>,
+    pub(crate) retry_policy: Box<dyn RetryPolicy>,
+    pub(crate) speculative_execution_policy: Option<Arc<dyn SpeculativeExecutionPolicy>>,
+}
+
+impl ExecutionProfile {
+    pub(crate) fn new_from_inner(inner: ExecutionProfileInner) -> Self {
+        Self(Arc::new(inner))
+    }
+
+    /// Creates a blank builder that can be used to construct new ExecutionProfile.
+    pub fn builder() -> ExecutionProfileBuilder {
+        ExecutionProfileBuilder {
+            request_timeout: None,
+            consistency: None,
+            serial_consistency: None,
+            load_balancing_policy: None,
+            retry_policy: None,
+            speculative_execution_policy: None,
+        }
+    }
+
+    /// Creates a builder having all options set to the same as set in this ExecutionProfile.
+    pub fn builder_derived(&self) -> ExecutionProfileBuilder {
+        ExecutionProfileBuilder {
+            request_timeout: Some(self.0.request_timeout),
+            consistency: Some(self.0.consistency),
+            serial_consistency: self.0.serial_consistency,
+            load_balancing_policy: Some(self.0.load_balancing_policy.clone()),
+            retry_policy: Some(self.0.retry_policy.clone()),
+            speculative_execution_policy: Some(self.0.speculative_execution_policy.clone()),
+        }
+    }
+
+    /// Returns a new handle to this ExecutionProfile.
+    pub fn handle(&self) -> ExecutionProfileHandle {
+        ExecutionProfileHandle(Arc::new(ArcSwap::new(self.0.clone())))
+    }
+}
+
+/// A handle that points to an ExecutionProfile. Its goal is to enable remapping
+/// all associated entities (query/Session) to another execution profile at once.
+#[derive(Debug, Clone)]
+pub struct ExecutionProfileHandle(Arc<ArcSwap<ExecutionProfileInner>>);
```

Comment:
Although I like the design of the handles better than the cpp driver's design of named execution profiles, I think it can be easy to confuse them and pass the wrong handle sometimes. In order to aid in debugging such issues I think we should allow users to add optional string labels to the "shared mutable execution profile" so that the users can quickly identify to which "smep" their handle points to. The label would be optional and, of course, immutable.

Reply:
IIUC, you propose adding an `Option<String> ` field to `ExecutionProfileHandle` ? Then, cloning a handle would clone that field as well.

Reply:
Looks like it is done in the way I wished for, so no need to improve it further here.